localstack/services/s3/v3/provider.py

```diff
@@ -0,0 +1,1231 @@
+        if not objects:
+            raise MalformedXML()
+
+        # TODO: max 1000 delete at once? test against AWS?
+        # TODO: implement ByPassGovernance
+        # TODO: implement Locking error
+
+        quiet = delete.get("Quiet", False)
+        deleted = []
+        errors = []
+
+        for to_delete_object in objects:
+            # TODO: beware of key encoding (XML?)
+            object_key = to_delete_object.get("Key")
+            version_id = to_delete_object.get("VersionId")
+            if s3_bucket.versioning_status is None:
+                if version_id and version_id != "null":
+                    errors.append(
+                        Error(
+                            Code="NoSuchVersion",
+                            Key=object_key,
+                            Message="The specified version does not exist.",
+                            VersionId=version_id,
+                        )
+                    )
+                    continue
+
+                found_object = s3_bucket.objects.pop(object_key, None)
+                if found_object:
+                    self._storage_backend.remove(bucket, found_object)
+
+                if not quiet:
+                    deleted.append(DeletedObject(Key=object_key))
+
+                continue
+
+            if not version_id:
+                delete_marker_id = generate_version_id(s3_bucket.versioning_status)
+                delete_marker = S3DeleteMarker(key=object_key, version_id=delete_marker_id)
+                s3_bucket.objects.set(object_key, delete_marker)
+                if not quiet:
+                    deleted.append(
+                        DeletedObject(
+                            DeleteMarker=True,
+                            DeleteMarkerVersionId=delete_marker_id,
+                            Key=object_key,
+                        )
+                    )
+                continue
+
+            if not (
+                found_object := s3_bucket.objects.pop(object_key=object_key, version_id=version_id)
+            ):
+                errors.append(
+                    Error(
+                        Code="NoSuchVersion",
+                        Key=object_key,
+                        Message="The specified version does not exist.",
+                        VersionId=version_id,
+                    )
+                )
+                continue
+
+            if not quiet:
+                deleted_object = DeletedObject(
+                    Key=object_key,
+                    VersionId=version_id,
+                )
+                if isinstance(found_object, S3DeleteMarker):
+                    deleted_object["DeleteMarker"] = True
+                    deleted_object["DeleteMarkerVersionId"] = found_object.version_id
+
+                deleted.append(deleted_object)
+
+            if isinstance(found_object, S3Object):
+                self._storage_backend.remove(bucket, found_object)
+
+        # TODO: request charged
```

Comment:
?

Reply:
is this comment still about batching delete?